### ISSUE TITLE
NLog does not function in a partially trusted environment

### ISSUE THREAD

**Issue by jonreis** [NLog/NLog#1494]
Bug
NLog version: 4.3.4 (NLog 2.1 works fine)
.Net 4.5, godaddy hosting platform

I am running a WCF service under a godaddy hosting service. When I attempt to upgrade the service to use NLog 4.3.4, my service will not start. I get the following exception:
- What is the current result?

An unhandled exception was generated during the execution of the current web request. Information regarding the origin and location of the exception can be identified using the exception stack trace below.

Stack Trace: 

```
[SecurityException: Request failed.]
   NLog.Common.InternalLogger.LogAssemblyVersion(Assembly assembly) +0
   NLog.LogFactory.LogConfigurationInitialized() +95
   NLog.LogFactory.get_Configuration() +637
   NLog.LogFactory.GetLogger(LoggerCacheKey cacheKey) +1149
   NLog.LogFactory.GetLogger(String name) +127
   NLog.LogManager.GetLogger(String name) +50
   HL7Spy.Service.MyServiceHost..cctor() +66

```

- What is the expected result?

I would expect that no exceptions "leak" of the NLog. My application/web service should still start up regardless of whether NLog can be configured, or not.

This is not a problem in NLog 2.1.0.0 which was the version prior to upgrading to the latest.

**Comment by 304NotModified** [comment]
Is the exception even thrown when throwExceptions = false

**Comment by jonreis** [comment]
Yes, that is true, it should be the default, and I tried setting

      LogManager.ThrowExceptions = false;

explicitly, and it still has the error.

**Comment by 304NotModified** [comment]
Is above the full exception posted? I expect some details after . "I get the following exception:"

**Comment by jonreis** [comment]
Hi Julian. That is all I get out of the webserver.

On a side note, if I add the marked code below to LogFactory. The problem goes away.

        private void LogConfigurationInitialized()
        {
          InternalLogger.Info("Configuration initialized.");
          if (AppDomain.CurrentDomain.IsFullyTrusted) // this is the new code
            InternalLogger.LogAssemblyVersion(typeof (ILogger).Assembly);          
        }


I am just reading up on why this works. I would have thought the try{}catch{} in LogAssemblyVersion would have sufficed, but for some reason the exception bubbles up from LogConfigurationInitialized().

**Comment by 304NotModified** [comment]
Any idea which call is raising the exception? And can't get SecurityException catched (there is a `catch`), see https://github.com/NLog/NLog/blob/04944f6637f5979627113bc6f1eed23de00784bc/src/NLog/Common/InternalLogger.cs#L356-L374

**Comment by jonreis** [comment]
Here is some more stack information:

[Exception: System.Security.SecurityException: Request failed.
   at NLog.Common.InternalLogger.LogAssemblyVersion(Assembly assembly)
   at NLog.LogFactory.LogConfigurationInitialized()
The action that failed was:
LinkDemand
The type of the first permission that failed was:
System.Security.PermissionSet
The Zone of the assembly that failed was:
MyComputer]
   NLog.LogFactory.LogConfigurationInitialized() +168
   NLog.LogFactory.get_Configuration() +637
   NLog.LogFactory.GetLogger(LoggerCacheKey cacheKey) +1149
   NLog.LogFactory.GetLogger(String name) +127
   NLog.LogManager.GetLogger(String name) +50
   HL7Spy.Service.MyServiceHost..cctor() +66

**Comment by 304NotModified** [comment]
don't get it. The `catch` should catch is. It's can be catched as seen in this example: https://msdn.microsoft.com/en-us/library/system.security.securityexception(v=vs.110).aspx

**Comment by 304NotModified** [comment]
or is it the 'assemby.Fullname`?

**Comment by jonreis** [comment]
It seems to be the way .NET walks the stack for permissions. See here: https://msdn.microsoft.com/en-us/library/c5tk9z76(v=vs.110).aspx

If you place the catch anywhere above LogAssemblyVersion(), it gets caught.

I placed it in the LogFactory.Configuration getter, since I don't think you want security exceptions to bubble up when its being initialized.

                            this.config.InitializeAll();
                            LogConfigurationInitialized();
                        } catch(SecurityException)
                        {  
                          // do nothing 
                        }

This seems to work when I test with godaddy.

There must be a way of replicating this in a test. I will see if I can figure that out.

**Comment by jonreis** [comment]
> or is it the assemby.Fullname in the catch?

No, I tried commenting out everything in the catch, and I still get the exception.

**Comment by 304NotModified** [comment]
FYI, I have no problems with the fix, but I like to understand it fully so I can recognize the issue in other parts.

**Comment by jonreis** [comment]
Me too. If you would like to do a shared meeting on skype, let me know I can demonstrate it in about 5 minutes.

**Comment by jonreis** [comment]
Just as an FYI, this is not an issue on godaddy anymore since one can set:

    <system.web>
        <trust level="Full" />
    </system.web>

I had 

    <system.web>
        <trust level="Medium" />
    </system.web>

from back when this was the only level you were allowed

**Comment by 304NotModified** [comment]
ah, well partially trust is also legacy AFAIK

**Comment by 304NotModified** [comment]
> Me too. If you would like to do a shared meeting on skype, let me know I can demonstrate it in about 5 minutes.

Thanks, but i'm a bit too busy for that ATM

**Comment by 304NotModified** [comment]
As medium trust is obsolete (http://stackoverflow.com/questions/16849801/is-trying-to-develop-for-medium-trust-a-lost-cause), we won't try to fix this.

**Comment by snakefoot** [comment]
Also reported by #1870, and fixed with  #1874
